user-console/src/main/resources/org/pentaho/mantle/public/browser/js/browser.js

```diff
@@ -1555,10 +1576,72 @@ define([
 
         if (elementBottom > scrollBottom) {
           listboxNode.scrollTop = elementBottom - listboxNode.clientHeight;
-        } else if (selectedOption.offsetTop < listboxNode.scrollTop) {
-          listboxNode.scrollTop = selectedOption.offsetTop;
+        } else if ((selectedOption.offsetTop - listboxNode.offsetTop) < listboxNode.scrollTop) {
+          listboxNode.scrollTop = selectedOption.offsetTop - listboxNode.offsetTop;
+        }
+
+      }
+    },
+
+    findItemToFocus: function (key, activeDescendant, listboxNode) {
+      var itemList = listboxNode.querySelectorAll('[role="option"]');
+      var character = String.fromCharCode(key);
+      var searchIndex = 0;
+
+      if (!this.model.get("keysSoFar")) {
+        for (var i = 0; i < itemList.length; i++) {
+          if (itemList[i].getAttribute('id') === activeDescendant) {
+            searchIndex = i;
+          }
         }
+      }
+      var keysSoFar = this.model.get("keysSoFar") + character;
+      this.model.set("keysSoFar", keysSoFar);
+      this.clearKeysSoFarAfterDelay(this.model.get("keysSoFar"));
 
+      var nextMatch = this.findMatchInRange(itemList, searchIndex + 1, itemList.length);
+      if (!nextMatch) {
+        nextMatch = this.findMatchInRange(itemList, 0, searchIndex);
+      }
+      return nextMatch;
+    },
+
+    findMatchInRange: function (list, startIndex, endIndex) {
+      // Find the first item starting with the keysSoFar substring, searching in
+      // the specified range of items
+      for (var n = startIndex; n < endIndex; n++) {
+        var label = list[n].innerText;
+        if (label && label.toUpperCase().indexOf(this.model.get("keysSoFar")) === 0) {
+          return list[n];
+        }
+      }
+      return null;
+    },
+
+    clearKeysSoFarAfterDelay: function () {
+      if (keyClear) {
+        clearTimeout(keyClear);
+        keyClear = null;
+      }
+      var keyClear = setTimeout(
```

Comment:
`keyClear` variable must instead be declared as a property of `this`. You can define it first in this class' `initialize` method (e.g. `this.keyClear = null`). Also, a better named would be, e.g., `keyClearTimeoutHandle`.